[Skia] Add support for notification icons

TingPing opened the pull request:
#### fc415ab5290aebf1c5d7c8948eb89f7661eecc9b
<pre>
[Skia] Add support for notification icons
<a href="https://bugs.webkit.org/show_bug.cgi?id=270755">https://bugs.webkit.org/show_bug.cgi?id=270755</a>

Reviewed by Michael Catanzaro.

* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
(WebKit::fdWStream::fdWStream):
(WebKit::IconCache::iconPath):

Canonical link: <a href="https://commits.webkit.org/275946@main">https://commits.webkit.org/275946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/apple/WebKit/commit/5bd968fbae1a6f3ddd89d4f8f3e754dc1e9d940a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22289 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45901 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39393 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19712 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35776 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43838 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37282 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16764 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1323 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39533 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38660 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47446 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14973 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42567 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19718 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41222 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19897 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5892 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19349 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -61,6 +67,48 @@ namespace WebKit {
+// Simple stream that writes to an fd.
+class fdWStream : public SkWStream {
+public:
+    fdWStream(int fd)
```
```suggestion
    explicit fdWStream(int fd)
```

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -114,8 +162,16 @@ class IconCache {
+            if (!surface->peekPixels(&pixmap) || !SkPngEncoder::Encode(&stream, pixmap, { })) {
+                g_warning("Failed to encode notification icon to PNG");
+                close(fd);
+                return CString();
```
Style nit:

```suggestion
                return { };
```

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -61,6 +67,48 @@ namespace WebKit {
+class fdWStream : public SkWStream {
+public:
+    fdWStream(int fd)
+        : m_fd(fd)
```
Give `fdWStream` a virtual destructor and close the fd there. i.e. let the stream object take ownership of the fd.

Or better: use a WTF::UnixFileDescriptor instead of an int. Then you don't even need to write the destructor. And the ownership will be clear. Something like this (untested):

```suggestion
        : m_fd(fd, UnixFileDescriptor::Adopt)
```

carlosgcampos commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -61,6 +67,48 @@ namespace WebKit {
+ssize_t (*writeToFD)(int, const void*, size_t) = write;
+
+// Simple stream that writes to an fd.
+class fdWStream : public SkWStream {
```
Wouldn't it be easier to just close the fd and use `SkFILEWStream` with the given path?

TingPing replied:
I don't think this class is very complex, and using a file path in tmp is usually not recommended since its racey. Maybe not a realistic problem but we have a valid FD why not use it.

mcatanzaro replied:
> Wouldn't it be easier to just close the fd and use SkFILEWStream with the given path?

Surely that's worse than what Patrick has done here.

carlosgcampos replied:
Can g_file_open_tmp return the same file path? I didn't say it's better, just that it's a lot easier. `FileSystem::createTemporaryFile()` does that, just closing the fd.

TingPing replied:
It is extremely unlikely, but `/tmp` is a shared dir for everything, so its recommended to just open a valid fd and hold it.

TingPing replied:
I can change it if you really think its easier. This class is just `write()` basically.

carlosgcampos replied:
No, I'm ok if you both insist it's better this way.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -61,6 +67,48 @@ namespace WebKit {
+ssize_t (*writeToFD)(int, const void*, size_t) = write;
+
+// Simple stream that writes to an fd.
+class fdWStream : public SkWStream {
```
The one other thing worth mentioning is that `fdWStream` doesn't follow our usual rules for naming classes (start with capital letter, avoid acronyms). I didn't mention this yesterday because the class is only used in one place, so I don't care very much. And style checker didn't complain. And I understand there is value in keeping the `WStream` suffix as that matches Skia. Still, you could consider `FileDescriptorWStream`.

TingPing replied:
:+1: Went with `FileDescriptorWriteStream`

TingPing committed:
[Skia] Add support for notification icons
https://bugs.webkit.org/show_bug.cgi?id=270755

Reviewed by Michael Catanzaro.

* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
(WebKit::fdWStream::fdWStream):
(WebKit::IconCache::iconPath):

Canonical link: https://commits.webkit.org/275946@main
